vars/cloudFoundryDeploy.groovy

```diff
@@ -232,11 +232,46 @@ void call(Map parameters = [:]) {
             .use()
 
         if (config.useGoStep == true) {
-            List credentials = [
-                [type: 'usernamePassword', id: 'cfCredentialsId', env: ['PIPER_username', 'PIPER_password']],
-                [type: 'usernamePassword', id: 'dockerCredentialsId', env: ['PIPER_dockerUsername', 'PIPER_dockerPassword']]
-            ]
-            piperExecuteBin(parameters, STEP_NAME, 'metadata/cloudFoundryDeploy.yaml', credentials)
+            def creds = []
+            if (config.mtaExtensionCredentials) {
+                config.mtaExtensionCredentials.each { key, credentialsId ->
+                    creds << string(credentialsId: credentialsId, variable: credentialsId)
+                }
+            }
+
+            withCredentials(creds) {
+                def credMapping = [:]
+                if (config.mtaExtensionCredentials) {
+
+                    // In the go layer we expect the string credentials as environment variables
+                    // hence we retrieve the values here inside Jenkins and set it as docker
+                    // environment variables so that they are available as expected as environment
+                    // variables.
+
+                    config.mtaExtensionCredentials.each { key, credentialsId ->
+                        credMapping[key] = this."${credentialsId}"
+                    }
+
+                    parameters = [:].plus(parameters)
+                    if (parameters.dockerEnvVars == null) {
+                        parameters.dockerEnvVars = [:]
+                    }
+                    parameters.dockerEnvVars << credMapping
+                }
+
+                List credentials = [
+                    [type: 'usernamePassword', id: 'cfCredentialsId', env: ['PIPER_username', 'PIPER_password']],
+                    [type: 'usernamePassword', id: 'dockerCredentialsId', env: ['PIPER_dockerUsername', 'PIPER_dockerPassword']]
+                ]
+                withEnv(toEnvironmentList(credMapping)) {
+                    /*
+                        We need the withEnv here in case we do not forward to docker inside piperExecuteBin.
+                        In this case we execute locally and in that case the credentials needs to be present
+                        in the current environment.
+                    */
+                    piperExecuteBin(parameters, STEP_NAME, 'metadata/cloudFoundryDeploy.yaml', credentials)
+                }
+            } // withCredentials ends here in order to avoid leaking the credentials somewhere during executing the piper go binary.
```

Comment:
Why can't we just put the additional credentials dynamically into the existing list "credentials"? Something like:
```groovy
List credentials = [
    [type: 'usernamePassword', id: 'cfCredentialsId', env: ['PIPER_username', 'PIPER_password']],
    [type: 'usernamePassword', id: 'dockerCredentialsId', env: ['PIPER_dockerUsername', 'PIPER_dockerPassword']]
]
if (config.mtaExtensionCredentials) {
    config.mtaExtensionCredentials.each { key, credentialsId ->
        credentials.add([type: 'secretText', id: "$credentialsId", env: ["$credentialsId"]])
    }
}
piperExecuteBin(...
```

Reply:
>Why can't we just put

To my understanding the resolved credentials variables are _groovy variables_, not _environment variables in a shell_. But I will check the details and report back if the approach outlined above would work.

Reply:
`withCredentials` provides apparently the corresponding variables as _groovy_ and as _shell_ variables. I was wrong above. As a side-learning it might not be required to have the `withEnv` block. I will check tomorrow.

But adding the credentials to the `credentials` list does not do the job anyway since inside the `piperExecuteBin` `credentialWrapper` function we have checks for the config (... `if (config[cred.id])`). Since the config does normally not contain corresponding IDs this does not work. See [here](https://github.com/SAP/jenkins-library/blob/master/vars/piperExecuteBin.groovy#L172). Now we can ask "what is the purpose of the `if (config[cred.id])`-check? But I guess there is a good reason ...

Reply:
I am never a fan of unnecessary code-duplication. piperExecuteBin already contains code to expose the values of certain credential IDs as environment variables to the go binary. If you need a little more flexibility, each entry in the credentials list passed to piperExecuteBin is itself a bunch of options (type, id, env-var-name, ...). So why not extend those options by the additional requirement, which is "config must configure the id -> true/false". Then you extend the existing check by the value from the credential entry, which defaults to "true", but in your use-case, you pass "false".

Reply:
I.e.
```groovy
if (config.mtaExtensionCredentials) {
    config.mtaExtensionCredentials.each { key, credentialsId ->
        credentials.add([type: 'secretText', id: "$credentialsId", env: ["$credentialsId"], checkConfig: false])
    }
}
```

and in piperExecuteBin:

```groovy
if (!cred.checkConfig || config[cred.id]) {
   ...
}
```

Reply:
> So why not extend those options

That we can also do.

Reply:
@stippi2 : I sketched the approach outlined by you above here: #2437.

I'm somehow reluctant with that approach. The coding inside `piperExecuteBin` wrt credentials handling gets less intuitive or - with other words - harder to understand when having the option to have an indirection into the config or not.

Of course the approach is technically possible. Nevertheless I personally think we should keep it simple. Hence IMO it would be better _not_ to merge that PR (#2437) and to keep the approach here in this PR with the additional `withCredentials`.

Not strongly against that approach here on the other hand. Maybe we should collect more opinions.

Beside that it was possible to shrink the approach here since the credentials are available through `withCredentials` as groovy variables and as environment variables - which is very nice.

Reply:
#2437 doesn't add a lot of complexity, IMH, I like it. You could also say the code has become a little cleaner in that the id is extracted from the config only in one place. We can discuss tomorrow in the sync.